### PR TITLE
feat(includers/openapi): request and response body params

### PR DESCRIPTION
**Pull request by amosov-f**
Tables for request and response bodies

Снимок экрана 2022-12-30 в 19 23 42

### CODE REVIEW

File: src/packages/credentials/yandex-oauth.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ async function getYandexOAuthTokenFromHomeDir() {
             throw new Error(error);
         }
     } catch (err) {
-        logger.error(error);
+        logger.error('', error);
```
Comment (amosov-f) [inline src/packages/credentials/yandex-oauth.ts:32]: was compilation error

**Review by moki — CHANGES_REQUESTED** [review]
at least minor changes that i mentioned needed

also please clean up commit history(local rebase with squash) with appropriate commit message(feat(includers/openapi): ... )

File: src/packages/credentials/yandex-oauth.ts (hunk 1 above)
Comment (amosov-f) [reply src/packages/credentials/yandex-oauth.ts]: reverted

File: src/services/includers/batteries/openapi/constants.ts [hunk 2]
```diff
@@ -7,6 +7,10 @@ const TAGS_SECTION_NAME = 'Sections';
 const ENDPOINTS_SECTION_NAME = 'Endpoints';
 const REQUEST_SECTION_NAME = 'Request';
 const PARAMETERS_SECTION_NAME = 'Parameters';
+const PATH_PARAMS_SECTION_NAME = 'Path params';
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/constants.ts]: fixed

File: src/services/includers/batteries/openapi/generators/common.ts [hunk 3]
```diff
@@ -45,6 +45,11 @@ function block(elements: any[]) {
     return elements.filter(Boolean).join(BLOCK);
 }
 
+// https://stackoverflow.com/a/49834158
+function escape(cellContent: string) {
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/common.ts]: fixed

File: src/services/includers/batteries/openapi/generators/endpoint.ts [hunk 4]
```diff
@@ -1,34 +1,40 @@
-import stringify from 'json-stringify-safe';
-
 import {block, title, body, table, code, cut} from './common';
 import {
-    DESCRIPTION_SECTION_NAME,
+    COOKIES_SECTION_NAME,
+    HEADERS_SECTION_NAME,
+    PATH_PARAMS_SECTION_NAME, QUERY_PARAMS_SECTION_NAME,
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/endpoint.ts]: fixed

File: src/services/includers/batteries/openapi/generators/endpoint.ts [hunk 5]
```diff
@@ -1,34 +1,40 @@
...
+import {prepareTableRowData, prepareSampleObject, tableFromSchema, tableParameterName} from './traverse';
 
-function endpoint(data: Endpoint) {
+function endpoint(allRefs: Refs, data: Endpoint) {
+    // try to remember, which tables we are already printed on page
+    const pagePrintedRefs = new Set<string>();
     const page = [
         title(1)(data.summary ?? data.id),
-        description(data.description),
+        data.description?.length && block([body(data.description)]),
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/endpoint.ts]: fixed

**Review by moki — CHANGES_REQUESTED** [review]

File: src/services/includers/batteries/openapi/generators/traverse.ts [hunk 6]
```diff
@@ -0,0 +1,195 @@
+import {Refs} from '../types';
+import {JSONSchema6} from 'json-schema';
+import {table} from './common';
+
+type TableRow = [string, string, string];
+
+function anchor(ref: string) {
+    return `<a href='#${ref.toLowerCase()}'>${ref}</a>`;
```
Comment (moki) [inline src/services/includers/batteries/openapi/generators/traverse.ts:8]: slugify href

https://github.com/yandex-cloud/yfm-docs/blob/c53392171534d7089a00605664815baf2243bcdb/src/services/includers/batteries/openapi/parsers.ts#L75

**Review by moki — CHANGES_REQUESTED** [review]

File: src/services/includers/batteries/openapi/generators/endpoint.ts [hunk 7]
```diff
@@ -39,37 +46,85 @@ function request(path: string, method: Method, servers: string[]) {
...
 
-function response(resp: Response) {
+function response(allRefs: Refs, visited: Set<string>, resp: Response) {
     return block([
-        title(3)(resp.code),
-        title(4)(DESCRIPTION_SECTION_NAME),
+        title(2)(resp.code),
         body(resp.description),
-        resp.schemas?.length && block(resp.schemas.map(schema)),
+        resp.schemas?.length && block(resp.schemas.map((s) => openapiBody(allRefs, visited, s))),
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/endpoint.ts:122]: fixed

File: src/services/includers/batteries/openapi/generators/traverse.ts [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+    const enumValues = value.enum?.map((s) => `\`${s}\``).join(', ');
+    if (enumValues) {
+        description += `<br>Enum: ${enumValues}`;
+    }
+    return {type: `${value.type}`, description};
+}
+
+// find dereferenced object from schema in all components/schemas
+function findRef(allRefs: Refs, value: JSONSchema6): string | undefined {
+    for (const [k, v] of Object.entries(allRefs)) {
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/traverse.ts:88]: flxed

File: src/services/includers/batteries/openapi/generators/traverse.ts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    refs: string[];
+};
+
+function prepareSchemaTable(refs: Refs, schema: JSONSchema6): PrepareSchemaTableResult {
+    const result: PrepareSchemaTableResult = {rows: [], refs: []};
+    const merged = merge(schema);
+    if (!merged) {
+        return result;
+    }
+    Object.entries(merged.properties || {}).forEach(([key, v]) => {
```
[the remark this answers is not in the document]
Comment (amosov-f) [reply src/services/includers/batteries/openapi/generators/traverse.ts:36]: fixed

**Review by moki — APPROVED** [review]

**Pull request merged**